text/0000-gen-fn.md

```diff
@@ -0,0 +1,476 @@
+
+## Error handling
+
+Within `#[rustc_gen]` blocks, the `?` operator desugars as follows.  When its
+argument returns a value indicating "do not short circuit"
+(e.g. `Option::Some(..)`, `Result::Ok(..)`, `ControlFlow::Continue(..)`), that
+value becomes the result of the expression as usual.  When its argument
+returns a value indicating that short-circuiting is desired
+(e.g. `Option::None`, `Result::Err(..)`, `ControlFlow::Break(..)`), the value
+is first yielded (after being converted by `From::from` as usual), then the
+block returns immediately.
+
+Even when `?` is used within a `#[rustc_gen]` block, the block must return a
+value of type unit or `!`.  That is, it does not return a value of `Some(..)`,
+`Ok(..)`, or `Continue(..)` as other such blocks might.
+
+However, note that when `?` is used within a `#[rustc_gen]` block, all `yield`
+statements will need to be given an argument of a compatible type.  For
+example, if `None?` is used in an expression, then all `yield` statements will
+need to be given arguments of type `Option`.
+
+## Fusing
+
+Like `Generators`, `Iterator`s produced by `#[rustc_gen]` panic when invoked again after they have returned `None` once.
+This will probably be fixed by special casing the generator impl if `Generator::Return = ()`, as we can trivially
+produce infinite values of the unit type.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## New keyword
+
+In the 2024 edition we reserve `gen` as a keyword. Previous editions will use `k#gen` to get the same features.
+
+## Error handling
+
+`foo?` in `#[rustc_gen]` blocks will stop iteration after the first error by desugaring to:
+
+```rust
+match foo.branch() {
+    ControlFlow::Break(err) => {
+        yield R::from_residual(err);
+        return;
+    },
+    ControlFlow::Continue(val) => val,
+}
+```
+
+This is the same behaviour that `collect::<Result<_, _>>()` performs
+on iterators over `Result`s.
+
+## Implementation
+
+This feature is mostly implemented via existing generators.
+We'll need additional desugarings and lots of work to get good diagnostics.
+
+### `#[rustc_gen] fn`
+
+`#[rustc_gen] fn` desugars to the function itself with the return type replaced by `impl Iterator<Item = $ret>` and its body wrapped in a `#[rustc_gen]` block.
+A `#[rustc_gen] fn`'s "return type" is its iterator's `yield` type.
+
+A `#[rustc_gen] fn` captures all lifetimes and #[rustc_gen]eric parameters into the `impl Iterator` return type (just like `async fn`).
+If more control over captures is needed, type alias impl trait can be used when it is stabilized.
+
+Like other uses of `impl Trait`, auto traits are revealed without being specified.
+
+### `#[rustc_gen]` blocks
+
+`#[rustc_gen]` blocks are the same as an unstable generator...
+
+* ...without arguments,
+* ...with an additional check forbidding holding borrows across `yield` points,
```

Comment:
Is there any appetite to consider this point in type-space between `Iterator` and `Generator`, where internal borrows are supported? It seems very limiting from my previous experiments to not be able to use `for x in &vec` at all.

Reply:
yes, imo that's what `Generator` should be, and our `Generator` could be renamed to `Coroutine` or sth. Figuring out lending iterators and self-referential iterators is something desirable, but with this RFC I'm trying to solve writing `Iterator`s and not closing the door on any of the future designs we'll want in this space. Once we have an incremental small step, doing the next one is easier than doing multiple at once.

Reply:
>  It seems very limiting from my previous experiments to not be able to use `for x in &vec` at all

Can someone clarify for me what this limitation is referring to? In what context the `for x in &vec` not usable? 

Are you e.g. saying that the `&Vec` is a parameter to the `#[rustc_gen] fn` and the `x` is being yielded, like so:

```rust
#[rustc_gen] fn demo<'a, X>(vec: &'a Vec<X>) -> &'a X {
     for x in vec { yield x; }
}
```


or do you mean that the `Vec` is a parameter to the `#[rustc_gen] fn` (and the `x` is still being yielded), like so:

```rust
#[rustc_gen] fn demo<X>(vec: Vec<X>) -> &'??? X {
    let borrowed = &vec;
    for x in borrowed { yield x; }
}
```

(The latter is problematic, at least at a surface level, because we do not have existential lifetimes. But I want to first figure out which case you two are talking about before I attempt to dissect it further.)

Reply:
The second case, but not attempting to yield a borrow, just use it internally to compute an owned value that gets yielded.

Reply:
(NOT A CONTRIBUTION)

There's two different issues:

1. A generator can't *store* references to its own state unless its pinned, but the `Iterator` interface doesn't take a pinned reference.
2. A generator can't *yield* references to its won state unless its interface borrows from itself, but the `Iterator` interface doesn't have the "lending" interface.

(The second issue also has the problem of having no clear way to express the lifetime, as you noted @pnkfelix.)

This RFC punts both issues, though probably at least the 1st issue needs to be figured out before you can stabilize something.

Reply:
From the implementation side the first issue is resolved by checking for `Unpin` just like `std::iter::from_generator` does. This will require additional work for good diagnostics, but the raw logic for it exists.

Reply:
That check is not necessary since the `#[rustc_gen]` transform is in control of the generator transform which is applied, it just needs to not set the flag that allows self-borrows (the difference between `|| yield;` and `static || yield;` generators). That has decent enough diagnostics already, they may just need the exact text changed depending on what this feature is called in documentation:

```
error[E0626]: borrow may still be in use when generator yields
6 |         for x in &values { yield *x; }
  |                  ^^^^^^^   -------- possible yield occurs here
```

The issue for this RFC is more: is there anything that might be designed differently to be forward-compatible with a syntax that allows self-borrows and generates a future `PinnedIterator` trait (or however it's spelled)?